cpp/arcticdb/codec/codec.hpp

```diff
@@ -48,59 +49,82 @@ void write_magic(Buffer& buffer, std::ptrdiff_t& pos) {
     pos += sizeof(MagicType);
 }
 
+/// @brief This should be the block data type descriptor when the shapes array is encoded as a block
+using ShapesBlockTDT = TypeDescriptorTag<DataTypeTag<DataType::INT64>, DimensionTag<Dimension::Dim0>>;
+/// @brief The type to be used for a block which represents shapes
+using ShapesBlock = TypedBlockData<ShapesBlockTDT>;
+
 //TODO  Provide tuning parameters to choose the encoding through encoding tags (also acting as config options)
-template<template<typename> class F, class TD>
+template<template<typename> class TypedBlock, class TD, EncodingVersion encoder_version>
 struct TypedBlockEncoderImpl;
 
-template<template<typename> class F, class TD>
-struct BlockEncoderHelper : public TypedBlockEncoderImpl<F, TD> {
-    using FieldType = F<TD>;
+template<template<typename> class F, class TD, EncodingVersion v>
+struct BlockEncoderHelper : public TypedBlockEncoderImpl<F, TD, v> {
+    using ValuesBlockType = F<TD>;
     using TypeDescriptorTag = TD;
     using DimTag = typename TD::DimensionTag;
     using DataTag = typename TD::DataTypeTag;
 };
 
-size_t encode_bitmap(
-    const util::BitMagic &sparse_map,
-    Buffer &out,
-    std::ptrdiff_t &pos);
+template<typename TD, EncodingVersion v>
+using BlockEncoder = BlockEncoderHelper<TypedBlockData, TD, v>;
 
-template<typename TD>
-using BlockEncoder = BlockEncoderHelper<TypedBlockData, TD>;
+/// @brief Options used by default to encode the shapes array of a column
+arcticdb::proto::encoding::VariantCodec shapes_encoding_opts();
 
-struct ColumnEncoder {
+/// @brief Utility class used to encode and compute the max encoding size for regular data columns for V1 encoding
+struct ColumnEncoder_v1 {
     static std::pair<size_t, size_t> max_compressed_size(
+        const arcticdb::proto::encoding::VariantCodec& codec_opts,
+        ColumnData& column_data);
+
+    static void encode(
         const arcticdb::proto::encoding::VariantCodec &codec_opts,
-        ColumnData &column_data);
+        ColumnData& column_data,
+        MutableVariantField variant_field,
```

Comment:
I don't really like typedefs that hide the attributes of a thing. If it's a reference to a mutable Thing then Think& tells me what I need to know where I read it, without having to go and look somewhere else to find that out.

Reply:
We already have this: https://github.com/man-group/ArcticDB/blob/60f55daac5e3efe6f951b718e3e9b706f9a47d58/cpp/arcticdb/codec/variant_encoded_field_collection.hpp#L14 however uses `const*` for both encoded fields. MutableVariantField is just the same but without the `const`.